java/ql/lib/semmle/code/java/regex/RegexTreeView.qll

```diff
@@ -0,0 +1,1031 @@
+      re.char_set_child(start, itemStart, result.getEnd())
+    )
+  }
+
+  override string getPrimaryQLClass() { result = "RegExpCharacterClass" }
+}
+
+/**
+ * A character range in a character class in a regular expression.
+ *
+ * Example:
+ *
+ * ```
+ * a-z
+ * ```
+ */
+class RegExpCharacterRange extends RegExpTerm, TRegExpCharacterRange {
+  int lower_end;
+  int upper_start;
+
+  RegExpCharacterRange() {
+    this = TRegExpCharacterRange(re, start, end) and
+    re.charRange(_, start, lower_end, upper_start, end)
+  }
+
+  /** Holds if this range goes from `lo` to `hi`, in effect is `lo-hi`. */
+  predicate isRange(string lo, string hi) {
+    lo = re.getText().substring(start, lower_end) and
+    hi = re.getText().substring(upper_start, end)
+  }
+
+  override RegExpTerm getChild(int i) {
+    i = 0 and
+    result.getRegex() = re and
+    result.getStart() = start and
+    result.getEnd() = lower_end
+    or
+    i = 1 and
+    result.getRegex() = re and
+    result.getStart() = upper_start and
+    result.getEnd() = end
+  }
+
+  override string getPrimaryQLClass() { result = "RegExpCharacterRange" }
+}
+
+/**
+ * A normal character in a regular expression, that is, a character
+ * without special meaning. This includes escaped characters.
+ *
+ * Examples:
+ * ```
+ * t
+ * \t
+ * ```
+ */
+class RegExpNormalChar extends RegExpTerm, TRegExpNormalChar {
+  RegExpNormalChar() { this = TRegExpNormalChar(re, start, end) }
+
+  /**
+   * Holds if this constant represents a valid Unicode character (as opposed
+   * to a surrogate code point that does not correspond to a character by itself.)
+   */
+  predicate isCharacter() { any() }
+
+  /** Gets the string representation of the char matched by this term. */
+  string getValue() { result = re.getText().substring(start, end) }
+
+  override RegExpTerm getChild(int i) { none() }
+
+  override string getPrimaryQLClass() { result = "RegExpNormalChar" }
+}
+
+/**
+ * A quoted sequence.
+ *
+ * Example:
+ * ```
+ * \Qabc\E
+ * ```
+ */
+class RegExpQuote extends RegExpTerm, TRegExpQuote {
+  string value;
+
+  RegExpQuote() {
+    exists(int inner_start, int inner_end |
+      this = TRegExpQuote(re, start, end) and
+      re.quote(start, end, inner_start, inner_end) and
+      value = re.getText().substring(inner_start, inner_end)
+    )
+  }
+
+  /** Gets the string matched by this quote term. */
+  string getValue() { result = value }
+
+  override string getPrimaryQLClass() { result = "RegExpQuote" }
+}
+
+/**
+ * A constant regular expression term, that is, a regular expression
+ * term matching a single string. This can be a single character or a quoted sequence.
+ *
+ * Example:
+ *
+ * ```
+ * a
+ * ```
+ */
+class RegExpConstant extends RegExpTerm {
+  string value;
+
+  RegExpConstant() {
+    (this = TRegExpNormalChar(re, start, end) or this = TRegExpQuote(re, start, end)) and
+    not this instanceof RegExpCharacterClassEscape and
+    // exclude chars in qualifiers
```

Comment:
Clarify why? Even in the context `a*`, the subexpression `a` seems like a constant? We also don't check for other syntactic parents that would make this non-constant -- `a|b`, `a(nonconst)` etc

Reply:
It looks like it actually exludes the `*` in `a*` from being considered a constant

Reply:
In fact, its purpose was to exclude characters in repitition quantifiers (`a{3}`) from being considered constants. It has now ben moved into the RegexString class.